fix(core): skip component ID collision warning during SSR

crisbeto opened the pull request:
The component ID collision check tries to account for components being replaced by checking for the `type`, however that might not work during SSR.

These changes disable the warning since it's primarily useful on the browser anyways.

Closes https://github.com/angular/angular-cli/issues/29408

alan-agius4 commented on packages/core/src/render3/definition.ts:
```diff
@@ -753,7 +753,12 @@ function getComponentId<T>(componentDef: ComponentDef<T>): string {
   if (typeof ngDevMode === 'undefined' || ngDevMode) {
```
```suggestion
  if (typeof ngDevMode === 'undefined' || ngDevMode && typeof ngServerMode === 'undefined' || !ngServerMode) {
```

alan-agius4 commented on packages/core/src/render3/definition.ts:
```diff
@@ -753,7 +753,12 @@ function getComponentId<T>(componentDef: ComponentDef<T>): string {
     if (GENERATED_COMP_IDS.has(compId)) {
       const previousCompDefType = GENERATED_COMP_IDS.get(compId)!;
-      if (previousCompDefType !== componentDef.type) {
+      const isNode = typeof process !== 'undefined' && !!process.on;
+
+      // Skip the check on the server since we can't guarantee the same component instance between
+      // requests. Note that we can't use DI to check if we're on the server, because the component
+      // hasn't been instantiated yet.
+      if (previousCompDefType !== componentDef.type && !isNode) {
```
```suggestion
      if (previousCompDefType !== componentDef.type) {
```

crisbeto committed:
fix(core): skip component ID collision warning during SSR

The component ID collision check tries to account for components being replaced by checking for the `type`, however that might not work during SSR.

These changes disable the warning since it's primarily useful on the browser anyways.

alan-agius4 reviewed (approved):
LGTM

ishbir commented:
Thanks for fixing this warning. I'd just like to add that from what I've seen with my code, it's not just an inconvenience. It seems to also break the functionality in the browser, and then I have to re-run ng serve to get the behavior. I am unfortunately not familiar enough with Angular to know whether it's due to this issue or another one, but I started seeing it occur with this warning.

alan-agius4 commented:
@ishbir, the IDs are mainly used for style encapsulation and hydration.

ishbir commented:
Is it possible that hydration is breaking somehow due to this? I do see breaks in actions performed in afterNextRender in the constructor which would suggest that the elements in the view do not get found/hydrated at the right time. I could try to create a minimally reproducible example if that would help.

alan-agius4 commented:
@ishbir, typically if there are hydration errors these are pointed out in the browser console. Please file an issue with a minimal reproduction.
